lib/storage/metadata/proxy/utils.js

```diff
@@ -0,0 +1,171 @@
+ * @return {object} ret - contains up to 4 string properties,
+ * @return {string} ret.namespace
+ * @return {string} ret.context
+ * @return {string} ret.bucketName
+ * @return {string} ret.objectName
+ */
+function getURIComponents(uri, logger) {
+    try {
+        if (uri.charAt(0) !== '/') {
+            return {};
+        }
+
+        const { pathname, query } = url.parse(uri);
+        const options = query ? querystring.parse(query) : {};
+        const typeIndex = pathname.indexOf('/', 1);
+        const bucketIndex = pathname.indexOf('/', typeIndex + 1);
+        const objectIndex = pathname.indexOf('/', bucketIndex + 1);
+
+        if (typeIndex === -1 || typeIndex === pathname.length - 1) {
+            return {};
+        }
+        if (bucketIndex === -1) {
+            return {
+                namespace: pathname.substring(1, typeIndex),
+                context: pathname.substring(typeIndex + 1),
+            };
+        }
+        if (bucketIndex === pathname.length - 1) {
+            return {
+                namespace: pathname.substring(1, typeIndex),
+                context: pathname.substring(typeIndex + 1, bucketIndex),
+            };
+        }
+        if (objectIndex === -1) {
+            return {
+                namespace: pathname.substring(1, typeIndex),
+                context: pathname.substring(typeIndex + 1, bucketIndex),
+                bucketName: pathname.substring(bucketIndex + 1),
+                options,
+            };
+        }
+        if (objectIndex === pathname.length - 1) {
+            return {
+                namespace: pathname.substring(1, typeIndex),
+                context: pathname.substring(typeIndex + 1, bucketIndex),
+                bucketName: pathname.substring(bucketIndex + 1, objectIndex),
+                options,
+            };
+        }
+        return {
+            namespace: pathname.substring(1, typeIndex),
+            context: pathname.substring(typeIndex + 1, bucketIndex),
+            bucketName: pathname.substring(bucketIndex + 1, objectIndex),
+            objectName: decodeURIComponent(pathname.substring(objectIndex + 1)),
+            options,
+        };
+    } catch (ex) {
+        logger.error('Invalid URI: failed to parse',
+                     { uri, error: ex, errorStack: ex.stack });
+        return null;
+    }
+}
+
+/**
+ * Extracts the body of the request through a callback
+ * @param {http.IncomingMessage} request - request received from bucketclient
+ * @param {Function} cb - function which has an interest in the request body.
+ *                        The first parameter is err and may be falsey
+ *                        The second parameter is the body of the request
+ * @return {undefined}
+ */
+function getRequestBody(request, cb) {
+    const body = [];
+    let bodyLen = 0;
+    request.on('data', data => {
+        body.push(data);
+        bodyLen += data.length;
+    }).on('error', cb).on('end', () => {
```

Comment:
`http.IncomingMessage` emits `close` if the client abruptly disconnects. We need to add a listener and destroy the stream/cleanup when the `close` event is emitted. This was the source of our TCP socket leak issue in January.